### PR TITLE
[WEBSITE-406] Add special extractors for GitHub and Confluence API

### PR DESCRIPTION
**Pull request by zbynek**
This PR adds a possibility to load documentation into the plugin site through APIs of Github and Confluence.

### CODE REVIEW

**Comment by zbynek** [comment]
restart build ...

**Comment by daniel-beck** [comment]
@zbynek Hi, thanks for the PR! As you can see I've wanted to review it from the moment you submitted it. Unfortunately it's been quite a while since I did anything with this component, and I don't think @michaelmccaskill is still around, so it might take a bit for me to find the time to remember how anything around here works.

File: README.adoc [hunk 1]
```diff
@@ -39,16 +39,18 @@ contains new data. If so the application will the reindex the Elasticsearch data
 == Run Local Plugin Site API
 
 ----
+GITHUB_CLIENT_ID="dummy"
```
Comment (rtyler) [inline README.adoc]: From an infrastructure standpoint, is this expected to be a personal access token in the future or should we be thinking about making a GitHub App with a client id, secret, etc for this?

Comment (zbynek) [reply README.adoc]: It's supposed to be a client id of a GitHub App, but the app does not need to have any callback URL, just the client ID and secret.

**Comment by zbynek** [comment]
> having a way to prioritize GitHub sources over Wikis for some plugins,

@oleg-nenashev what do you mean? Each plugin defines one URL for documentation and one for source code. If you want to use GH for documentation you can just point both URLs to your repo.

One thing I'm not sure about is how to pass the GitHub  app credentials to the plugin site. It is read from environment variables,  but I guess those should be set in Jenkinsfile. Not sure if/how I can do that or someone with access to Jenkins infrastructure should do it.

**Comment by oleg-nenashev** [comment]
> @oleg-nenashev what do you mean? Each plugin defines one URL for documentation and one for source code. If you want to use GH for documentation you can just point both URLs to your repo.

That is the part I am missing. Studying the code right now.

> One thing I'm not sure about is how to pass the GitHub app credentials to the plugin site. It is read from environment variables, but I guess those should be set in Jenkinsfile. Not sure if/how I can do that or someone with access to Jenkins infrastructure should do it.

We will need to add something for it, CC @olblak

**Comment by oleg-nenashev** [comment]
Some manual testing updates here...

* Plugin documentation information is stored in https://updates.jenkins.io/current/plugin-documentation-urls.json
* Some plugins already point to GitHub, e.g. the new `folder-auth` plugin
* With this patch, the backend still generates the Wiki source. I will debug the code to understand why

```json
{
   "firstRelease":"2019-08-11T06:25:55.50Z",
   "buildDate":"2019-08-11",
   "categories":[

   ],
   "dependencies":[
      {
         "name":"configuration-as-code",
         "title":"Configuration as Code",
         "optional":true,
         "version":"1.25",
         "implied":false
      },
      {
         "name":"cloudbees-folder",
         "title":"Folders",
         "optional":false,
         "version":"6.4",
         "implied":false
      },
      {
         "name":"trilead-api",
         "title":"Trilead API",
         "optional":false,
         "version":"1.0.4",
         "implied":true
      }
   ],
   "maintainers":[
      {
         "id":"AbhyudayaSharma",
         "name":"Abhyudaya Sharma",
         "email":"sharmaabhyudaya@gmail.com"
      }
   ],
   "excerpt":"Allows configuring users&#39; permissions to projects organized in folders from the Cloudbees Folder plugin.",
   "gav":"io.jenkins.plugins:folder-auth:1.0.1",
   "labels":[

   ],
   "name":"folder-auth",
   "previousTimestamp":"2019-08-11T06:25:55.00Z",
   "previousVersion":"1.0",
   "releaseTimestamp":"2019-08-11T06:56:54.00Z",
   "requiredCore":"2.164.1",
   "scm":{
      "issues":null,
      "link":"https://github.com/jenkinsci/folder-auth-plugin",
      "inLatestRelease":null,
      "sinceLatestRelease":null,
      "pullRequests":null
   },
   "sha1":"qX/d5QZ0NemyYP/ZY4mECQKozvI=",
   "stats":{
      "installations":[
         {
            "timestamp":1561953600000,
            "total":1
         }
      ],
      "installationsPercentage":[
         {
            "timestamp":1561953600000,
            "percentage":3.7600204545112725E-4
         }
      ],
      "installationsPerVersion":[
         {
            "version":"0.1-alpha-3-SNAPSHOT (private)",
            "total":1
         }
      ],
      "installationsPercentagePerVersion":[
         {
            "version":"0.1-alpha-3-SNAPSHOT (private)",
            "percentage":3.7600204545112725E-4
         }
      ],
      "currentInstalls":1,
      "trend":0
   },
   "title":"Folder-based Authorization Strategy",
   "url":"http://updates.jenkins-ci.org/download/plugins/folder-auth/1.0.1/folder-auth.hpi",
   "version":"1.0.1",
   "securityWarnings":null,
   "wiki":{
      "content":"<div>\n Documentation for this plugin is here: \n <a href=\"https://github.com/jenkinsci/folder-auth-plugin\">https://github.com/jenkinsci/folder-auth-plugin</a>\n</div>",
      "url":"https://github.com/jenkinsci/folder-auth-plugin"
   }
}
```

**Comment by oleg-nenashev** [comment]
PEBKAC. With GitHub client ID it works like a charm. Created https://issues.jenkins-ci.org/browse/INFRA-2228 to expose it inside the service. 

The only problem is with including images and relative links. All of that gets returned incorrectly from API:

```
Three types of roles are supported 'Global Roles', 'Folder Roles' and 'Agent Roles'.</p> \n  <p><a target=\"_blank\" rel=\"noopener noreferrer\" href=\"https://github.com/docs/images/screenshot.png\"><img src=\"https://github.com/docs/images/screenshot.png\" alt=\"Screenshot of the plugin\" style=\"max-width:100%;\"></a></p> \n  <p>To learn more about how to use this plugin, read the <a href=\"https://github.com/docs/usage.md\">docs</a>
```

@zbynek Any chance we could resolve links properly in the parser?

Resulting output:

image

**Comment by oleg-nenashev** [comment]
NVM, I am looking into the code to see why the existing replacer did not work for me

**Comment by casz** [comment]
Dno if anybody else found it funny.

WEBSITE-406.
406 is error code for Not Acceptable.
So Not Acceptable that plugin site cannot pull data from GitHub 🙊

**Comment by oleg-nenashev** [comment]
I will merge this change once we discuss https://issues.jenkins-ci.org/browse/INFRA-2228 with @olblak and egree on the implementation

**Comment by oleg-nenashev** [comment]
https://github.com/jenkins-infra/jenkins-infra/pull/1351 for the Infra update

**Pull request merged**